---
layout: vanilla
permalink: /flow/
description: "Documentation for Monocle."
modified: 2013-09-11
tags: [monocle, install, installation]
---
<!DOCTYPE html>
<html lang= "{{ page.lang | default: site.lang | default: "en" }}">

  {% include head.html %}

  <!-- See here for how to use the TOC control:
  https://afeld.github.io/bootstrap-toc/ -->
  <body  data-spy= "scroll" data-target= "#toc">
    {% include header.html %}

    <div class= "container">
      <div class= "row">
        <!-- sidebar, which will move to the top on a small screen -->
        <div class= "col-sm-4">
          <nav id= "toc" data-spy= "affix" data-toggle= "toc"></nav>
        </div>

<h2><p class="text-center">Workflow Steps at a Glance</p></h2>
<p class="text-center">Below, you can see snippets of code that highlight the main steps of Monocle.
Clink on the section headers to jump to the detailed sections describing each one. </p>

<br>

<h3><a href= "{{ site.url }}/docs/#getting-started-with-monocle">Store Data in a CellDataSet Object</a></h3>
The first step in working with Monocle is to load up your data into Monocle's
main class, <code>CellDataSet</code>:

<br>

{% highlight R %}
pd <- new("AnnotatedDataFrame", data = sample_sheet)
fd <- new("AnnotatedDataFrame", data = gene_annotation)
cds <- newCellDataSet(expr_matrix, phenoData = pd, featureData = fd)
{% endhighlight %}

<br>

<h3><a href= "{{ site.url }}/docs/#classifying-and-counting-cells">Classify cells with known marker genes</a></h3>
Next, leverage your knowledge of key marker genes to quickly and easily classify your cells by type:
{% highlight R %}
cth <- newCellTypeHierarchy()

MYF5_id <- row.names(subset(fData(cds), gene_short_name == "MYF5"))
ANPEP_id <- row.names(subset(fData(cds), gene_short_name == "ANPEP"))

cth <- addCellType(cth, "Myoblast", classify_func = function(x) { x[MYF5_id,] >= 1 })
cth <- addCellType(cth, "Fibroblast",
                   classify_func = function(x) { x[MYF5_id,] < 1 & x[ANPEP_id,] > 1 } )

cds <- classifyCells(cds, cth, 0.1)
{% endhighlight %}

<br>

<h3><a href= "{{ site.url }}/docs/#clustering-cells">Cluster your cells</a></h3>
You can easily cluster your cells to find new types:
{% highlight R %}
cds <- clusterCells(cds)
{% endhighlight %}

<br>

<h3><a href= "{{ site.url }}/docs/#constructing-single-cell-trajectories">Order cells in pseudotime along a trajectory</a></h3>
Now, put your cells in order by how much progress they've made through whatever process you're studying, such as differentiation,
reprogramming, or an immune response.

{% highlight R %}
disp_table <- dispersionTable(cds)
ordering_genes <- subset(disp_table, mean_expression >= 0.1)
cds <- setOrderingFilter(cds, ordering_genes)
cds <- reduceDimension(cds)
cds <- orderCells(cds)
{% endhighlight %}

<br>

<h3><a href= "{{ site.url }}/docs/#differential-expression-analysis">Perform differential expression analysis</a></h3>
Compare groups of cells in myriad ways to find differentially expressed genes,
controlling for batch effects and treatments as you like:
{% highlight R %}
diff_test_res <- differentialGeneTest(cds, fullModelFormulaStr = "~Media")
sig_genes <- subset(diff_test_res, qval < 0.1)
{% endhighlight %}
