---
layout: default
title: trapnell lab
---
<div class="container">

<h2> Install quick-start </h2>

<h3> Required software </h3>
Monocle runs in the <a href="http://www.r-project.org/">R statistical computing environment</a>. You will need R version 3.4 or higher, Bioconductor version 3.5, and monocle 2.4.0 or higher to have access to the latest features. To install <a href="http://bioconductor.org/install/">Bioconductor</a>:

{% highlight R %}
source("http://bioconductor.org/biocLite.R")
biocLite()
{% endhighlight %}

Once you've installed Bioconductor, you're ready to install Monocle and all of its required dependencies:

{% highlight R %}
> biocLite("monocle")
{% endhighlight %}

<h3> Testing the installation </h3>
To ensure that Monocle was installed correctly, start a new R session and type:

{% highlight R %}
library(monocle)
{% endhighlight %}

<h3> Installing Monocle 2 </h3>
Monocle 2 is currently available through GitHub. Once it works through Bioconductor's development-to-release cycle, which takes 6 months, it will become available via the steps above. To install it through GitHub, enter the following commands at the R console:

{% highlight R %}
install.packages("devtools")
devtools::install_github("cole-trapnell-lab/monocle-release@monocle2")
{% endhighlight %}

Monocle 2 has a number of new dependencies that Monocle 1 didn't. You may see errors when you try the above command. You can install the packages in the error message by typing (for example):

{% highlight R %}
biocLite(c("DDRTree", "pheatmap"))
{% endhighlight %}

If you install Monocle 2, make sure to have a look at the new [vignette]({{site.url}}/docs/) and [reference manual]({{site.url}}/docs/index.html), as many things have changed.



<h2> Analyzing data with Monocle </h2>

Monocle provides a number of tools you can use to analyze your single cell expression experiments. To get started, we must create a CellDataSet object. You can do this with the commands below:

{% highlight R %}
library(monocle)
sample_sheet <- read.delim("sc_expr_out/samples.table", row.names=1)
gene_annotations <- read.delim("sc_expr_out/genes.attr_table", row.names=1)
fpkm_matrix <- read.delim("sc_expr_out/genes.fpkm_table", row.names=1)
pd <- new("AnnotatedDataFrame", data = sample_sheet)
my_data <- newCellDataSet(exprs = as.matrix(fpkm_matrix), phenoData = pd, featureData = fd)
{% endhighlight %}

Now, you have created an object named "my_data" that stores your single-cell expression data. This object is the central object in Monocle. You will use it to identify differentially expressed genes and perform other analyses. To see what Monocle can do for you and how to proceed, please have a look at the [vignette (PDF)](http://www.bioconductor.org/packages/devel/bioc/vignettes/monocle/inst/doc/monocle-vignette.pdf). Good luck!
</div>
