<!DOCTYPE html>
<html lang= "en">

  <!-- <head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  <title>Monocle</title>
  <meta name="description" content="Monocle - powerful software toolkit for single-cell analysis">

  <link rel="stylesheet" href="/css/main.css">
  <link rel="canonical" href="http://localhost:4000/tutorials/">
  <link rel="alternate" type="application/rss+xml" title="Monocle" href="/feed.xml">

  
</head> -->

<head>
   <meta charset="utf-8">
   <meta http-equiv="X-UA-Compatible" content="IE=edge">
   <meta name="viewport" content="width=device-width, initial-scale=1">
   <!-- The above 3 meta tags *must* come first in the head; any other head content must come *after* these tags -->
   <title>Monocle</title>
   <meta name="description" content="Monocle - powerful software toolkit for single-cell analysis">
   <!-- <link rel="icon" href="../../favicon.ico"> -->

   <title>Theme Template for Bootstrap</title>

  <link rel="stylesheet" href="/css/main.css">
  <link rel="stylesheet" href="http://localhost:4000/node_modules/prismjs/themes/base16-atelierlakeside.light.css" />
  <link rel="stylesheet" href="https://cdn.rawgit.com/afeld/bootstrap-toc/v0.4.1/dist/bootstrap-toc.min.css">
  <!-- add after bootstrap.min.js -->
  <link rel="canonical" href="http://localhost:4000/tutorials/">


   <!-- HTML5 shim and Respond.js for IE8 support of HTML5 elements and media queries -->
   <!--[if lt IE 9]>
     <script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
     <script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
   <![endif]-->

   <!--  -->
 </head>


  <body  data-spy= "scroll" data-target= "#toc">
    <!-- <header class="site-header" role="banner"> -->
  <div class="container-fullwidth">
  <nav class="navbar navbar-inverse navbar-static-top">
    <div class="container">
      <div class="navbar-header">
        <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false" aria-controls="navbar">
          <span class="sr-only">Toggle navigation</span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="http://localhost:4000/">Monocle</a>
      </div>
      <div id="navbar" class="collapse navbar-collapse">
        <ul class="nav navbar-nav">
          <!-- <li class="active"><a href="#">Home</a></li> -->
          <li><a href="http://localhost:4000/docs/">Documentation</a></li>
          <!-- <li><a href="/features/">Features</a></li> -->
          <!-- <li><a href="/getting-started/">Getting started</a></li> -->

          <li><a href="http://localhost:4000/papers/">Publications</a></li>
          <li><a href="http://localhost:4000/tutorials/">Tutorials</a></li>
          <li><a href="https://github.com/cole-trapnell-lab/monocle-release">GitHub</a></li>
        </ul>
      </div><!--/.nav-collapse -->
    </div>
  </nav>
  </div>
  <!-- <div class="wrapper">
    
    
    <a class="site-title" href="/">Monocle</a>

    
      <nav class="site-nav">
        <input type="checkbox" id="nav-trigger" class="nav-trigger" />
        <label for="nav-trigger">
          <span class="menu-icon">
            <svg viewBox="0 0 18 15" width="18px" height="15px">
              <path fill="#424242" d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.031C17.335,0,18,0.665,18,1.484L18,1.484z"/>
              <path fill="#424242" d="M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0c0-0.82,0.665-1.484,1.484-1.484 h15.031C17.335,6.031,18,6.696,18,7.516L18,7.516z"/>
              <path fill="#424242" d="M18,13.516C18,14.335,17.335,15,16.516,15H1.484C0.665,15,0,14.335,0,13.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.031C17.335,12.031,18,12.696,18,13.516L18,13.516z"/>
            </svg>
          </span>
        </label>

        <div class="trigger">
          
            
            
          
            
            
          
            
            
          
            
            
          
            
            
          
            
            
          
            
            
          
            
            
          
            
            
          
            
            
          
            
            
          
            
            
          
            
            
          
            
            
          
            
            
          
            
            
          
            
            
          
        </div>
      </nav>
    
  </div> -->
</header>


    <div class= "container">
      <div class= "row">
        <!-- sidebar, which will move to the top on a small screen -->
        <div class= "col-sm-4">
          <nav id= "toc" data-spy= "affix" data-toggle= "toc"></nav>
        </div>
        <!-- main content area -->
        <div class= "col-sm-12">

          <div class="text-center">
            <h3 data-toc-skip>Below are a few examples on how to use Monocle to analyze particular datasets</h3>
          </div>

          <hr class="featurette-divider">




          <div class="row featurette">
            <div class="col-md-4">
              <img class="featurette-image img-responsive center-block" src="http://localhost:4000/images/olsson-complex-plot.png" alt="Monocle 1" height = 400/>
            </div>
            <div class="col-md-8">

              <h3 data-toc-skip>Olsson Dataset Analysis</h3>
              <br>
              <p> In this tutorial, we demonstrate how to use reversed graph embedding (RGE) in Monocle 2 to resolve the complicated haematopoiesis process which involves two major branch points. In regards to the branch points, HSC bifurcates into either megakaryote/erythroid progenitor (MEP) or granulocyte/monocyte progenitor (GMP) and the bifurcation from GMP into either granulocyte and monocyte progenitor.
                The reconstructed developmental trajectory is learned in four dimensions but can be visualized in two dimensions. With the trajectory learned, we are able to identify genes showing significant a bifurcation pattern during each lineage bifurcation through <a href = "http://localhost:4000/papers/">BEAM</a>. We can further combine ChIP-seq datasets and Irf8 (monocyte master regulator) for Gfi1 (granulocyte master regulator) as well as the transcription factors to build a regulatory network from master regulators to other important downstream regulators and then those regulators to their downstream targets. This network hierarchy is promising to resolve the fundamental regulatory mechanism for the hematopoiesis. We also provided a multi-way heatmap and multi-way kinetic curves to visualize important marker genes over the differentiation process. Some additional analyses are included, which are used for the tutorial on analyzing the Paul dataset.

                <div class="text-center">
                  <a class="btn btn-default btn-sm" href="http://localhost:4000/Olsson_dataset_analysis_final.html" role="button">HTML</a>
                  <a class="btn btn-default btn-sm" href="https://github.com/cole-trapnell-lab/monocle2-rge-paper/blob/master/Olsson_dataset_analysis_final.ipynb" role="button">ipynb</a>
                </div>


            </div>
          </div>

          <br>

          <hr class="featurette-divider">

          <div class="row featurette">
            <div class="col-md-4">
                <img class="featurette-image img-responsive center-block" src="http://localhost:4000/images/paul-a-c.png" alt="Monocle 1" height = 400/>
            </div>
            <div class="col-md-8">
              <h3 data-toc-skip>Paul Dataset Analysis</h3>
              <br>
              <p> In this tutorial, we demonstrate how to use Monocle 2 to resolve the complicated haematopoiesis process that contains five branch points by applying reversed graph embedding (RGE). To demonstrate this we will be analyzing the Paul dataset. The reconstructed developmental trajectory is learned in ten dimensions, selected based on PC variance explained, but can be visualized in two dimensions using a tree layout. We show that the structure of the tree we reconstructed matches well with the cell type assignment from the  <a href = "http://www.cell.com/cell/abstract/S0092-8674(15)01493-2">original paper</a>. When we use genes based on the Olsson dataset to define stemness or lineage specificity for each cell on the tree, we can see the smooth transition of cell differentation. Multi-way kinetic curves and multi-way heatmaps visualize the dynamics of six different cell fate commitments.

                <div class="text-center">
                  <a class="btn btn-default btn-sm" href="http://localhost:4000/Paul_dataset_analysis_final.html" role="button">HTML</a>
                  <a class="btn btn-default btn-sm" href="https://github.com/cole-trapnell-lab/monocle2-rge-paper/blob/master/Paul_dataset_analysis_final.ipynb" role="button">ipynb</a>
                </div>

            </div>
          </div>

          <br>

