<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <title>Monocle - Differential expression and time-series analysis for single-cell RNA-Seq and qPCR experiments</title>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
	<meta name="description" content="Open source tools for transcript assembly, differential expression, and
        differential regulation for RNA-Seq">
    <link rel="stylesheet" type="text/css" href="css/style.css"
      media="screen">
  </head>
  <body>
    <div id="wrap">
      <div id="top">
        <div class="lefts">
          <table cellpadding="2" width="100%">
            <tbody>
              <tr>
                <td> <a href="./index.html">
                    <h1>Monocle</h1>
                  </a>
                  <h2>Differential expression and time-series analysis for single-cell RNA-Seq and qPCR experiments</h2>
                </td>
                <td align="right" valign="middle"> 
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
      <div id="subheader">
        <table width="100%">
          <tbody>
            <tr>
              <td> <strong>Monocle</strong> is a toolkit for analyzing single-cell gene expression experiments.  Monocle was designed for RNA-Seq, but can also work with single cell qPCR. It performs differential expression analysis, and can find genes that differ between cell types or between cell states. When used to study an ongoing biological process such as cell differentiation, Monocle learns that process and places cells in order according to their progress through it. Monocle finds genes that are dynamically regulated during that process. <br>
                <br>
                Monocle was written by Cole Trapnell with input from Davide Cacchiarelli. <br>
                <br>
                Monocle is provided under the OSI-approved <a
                  href="http://en.wikipedia.org/wiki/Artistic_license">Artistic License (version 2.0)</a> </td>
              <td align="right" valign="middle"> <a
                  href="http://opensource.org"><img alt="Open Source
                    Software" src="images/osi-certified.gif" border="0"></a>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <div id="main">
        <div id="rightside">
	
          <h2>Site Map</h2>
          <div class="box">
            <ul>
              <li><a href="index.html">Home</a></li>
              <li><a href="tutorial.html">Getting started</a></li>
              <li><a href="monocle-manual.pdf">Manual (PDF)</a></li>
			  <li><a href="monocle.pdf">Vignette (PDF)</a></li>
			  <li><a href="igenomes.html">Index and annotation downloads</a></li>
			  <li><a href="protocol.html">Protocol</a></li>
            </ul>
          </div>

          <h2><u>News and updates</u></h2>
          <div class="box">
            <ul>
              <table width="100%">
                <tbody>
                  <tr>
                    <td>New releases and related tools will be announced
                      through the <a
                        href="https://lists.sourceforge.net/lists/listinfo/bowtie-bio-announce"><b>mailing
                          list</b></a></td>
                  </tr>
                </tbody>
              </table>
            </ul>
          </div>
          <h2><u>Getting Help</u></h2>
          <div class="box">
            <ul>
              <table width="100%">
                <tbody>
                  <tr>
                    <td>Questions about Monocle should be posted on our <a href="https://groups.google.com/forum/#!forum/monocle-users"><b>Google Group</b></a>. Please use <a
                        href="mailto:monocle.users@gmail.com">monocle.users@gmail.com</a> for private communications only.
                      Please do not email technical questions to
                      Monocle contributors directly.</td>
                  </tr>
                </tbody>
              </table>
            </ul>
          </div>
       
          <a href="./downloads">
            <h2><u>Releases</u></h2>
          </a>
          <div class="box">
            <ul>
              <table width="100%">
                <tbody>
                  <tr>
                    <td>version XXX</td>
                    <td align="right">XXX</td>
                  </tr>
                  <tr>
                    <td><a href="./downloads/cufflinks-2.1.1.tar.gz"
                        onclick="javascript:
                        pageTracker._trackPageview('/downloads/cufflinks_source');
                        ">&nbsp;&nbsp;&nbsp;Source code</a></td>
                  </tr>
                  <tr>
                    <td><a
                        href="./downloads/cufflinks-2.1.1.Linux_x86_64.tar.gz"
                        onclick="javascript:
                        pageTracker._trackPageview('/downloads/cufflinks');
                        ">&nbsp;&nbsp;&nbsp;Linux x86_64 binary</a></td>
                  </tr>
                  <tr>
                    <td><a
                        href="./downloads/cufflinks-2.1.1.OSX_x86_64.tar.gz"
                        onclick="javascript:
                        pageTracker._trackPageview('/downloads/cufflinks');
                        ">&nbsp;&nbsp;&nbsp;Mac OS X x86_64 binary</a></td>
                  </tr>
                </tbody>
              </table>
            </ul>
          </div>

          <h2>Related Tools</h2>
          <div class="box">
            <ul>
				<li><a href="http://cufflinks.cbcb.umd.edu">Cufflinks</a>: 
					Transcript assembly and differential expression for RNA-Seq
				<li><a href="http://compbio.mit.edu/cummeRbund/">CummeRbund</a>:
	                Visualization of RNA-Seq differential analysis</li>
              <li><a href="http://tophat.cbcb.umd.edu/">TopHat</a>:
                Alignment of short RNA-Seq reads</li>
              <li><a href="http://bowtie.cbcb.umd.edu">Bowtie</a>:
                Ultrafast short read alignment</li>
            </ul>
          </div>

          <h2>Publications</h2>
          <div class="box">
            <ul>
              <li style="font-size: x-small; line-height: 130%">
                <p>Trapnell C, Cacchiarelli D, Grimsby J, Pokharel P, Li S, Morse M, Lennon NJ, Livak KJ, Mikkelsen TS, Rinn JL<b> <a
                      href="http://XXX">the dynamics and regulators of cell fate decisions are revealed by pseudotemporal ordering of single cells</a></b> <br>
                  <i><a href="http://www.nature.com/nbt">Nature
                      Biotechnology</a></i> XXX</p>
                <br>
              </li>
            </ul>
          </div>
          <h2>Contributors</h2>
          <div class="box">
            <ul>
              <li><a href="http://www.cs.umd.edu/%7Ecole/">Cole Trapnell</a></li>
              <li>Davide Cacchiarelli</a></li>
              <li>Jonna Grimsby</li>
			  <li>Prapti Pokharel</li>
			  <li>Shuqiang Li</li>
			  <li>Michael Morse</li>
			  <li>Niall J. Lennon</li>
			  <li>Kenneth J. Livak</li>
			  <li>Tarjei S. Mikkelsen</li>
			  <li>John L. Rinn</li>
            </ul>
          </div>
        </div>
        <!-- End of "rightside" -->
		
        <div id="leftside">
          <table>
            <tbody>
              <tr>
                <td>
                  <h1> Getting started</h1>
                  <br/>
                  <div id="toc">
                    <ul>
                        <li><a href="#install">Installing Monocle</a></li>
                        <li><a href="#expression">Computing expression values for single cells</a></li>
						<li><a href="#monocle">Analyzing data with Monocle</a></li>
                    </ul>
                    <br/>
                  </div>
                  <br/>
                  <h2 id="inst">Install quick-start</h2>
                  <br/>
                  <strong>Required software</strong><br/>
                  <p>Monocle runs in the <a href="http://www.r-project.org/">R statistical computing environment</a>.  You will need R version 3.0 or higher.  You will need to install the following packages through CRAN:
				  <p>  <ul>
						<li>VGAM
						<li>irlba
						<li>matrixStats
						<li>igraph (version >= 0.7.0)
						<li>combinat
						<li>fastICA
						<li>grid
						<li>ggplot2
						<li>reshape2
						<li>plyr
						<li>parallel
						<li>methods
					  </ul>
  					<p>You can install these packages by starting an R session and typing:
					<br/>
  					<br/>  
  					<blockquote>
  					> install.packages(c("VGAM", "irlba", "matrixStats", "igraph",  <br>
									   "combinat", "fastICA", "grid", "ggplot2",  <br>
									   "reshape2", "plyr", "parallel", "methods"))
  					</blockquote>
					<br/> 
					You will also need to install <a href="http://bioconductor.org/install/">Bioconductor</a>: 
					<br/>
					<br/>
					<blockquote>
					> source("http://bioconductor.org/biocLite.R")  <br>
					> biocLite()
					</blockquote>
					
                  <br/>
				  <br/>
                  <strong>Installing a pre-compiled binary release</strong><br/>
                  <p>In order to make it easy to install Monocle, we
                    provide a few binary packages to save users from
                    occasionally frustrating process of building
                    Monocle. To use the binary packages, simply
                    download the appropriate one for your machine, change to the directory in which you saved it, and type: .</p>
					<br/>
					<br/>
					<blockquote>
					$ R CMD INSTALL monocle_0.99.0.tar.gz  <br>
					</blockquote>
                  <br/>
                  <strong>Installing Monocle from source</strong><br/>
                  <p>You can also build Monocle from the source code if you prefer. To do so, download the source tarball, change to the directory in which you saved it, and type</p>
					<br/>
					<br/>
					<blockquote>
					$ R CMD INSTALL monocle_0.99.0.tar.gz  <br>
					</blockquote>
                  <br/>

                  <strong>Testing the installation</strong><br/>
                  <p>To ensure that Monocle was installed correctly, start a new R session and type:</p>
				  <br/>
				  <blockquote>
				  > library(monocle)
				  </blockquote>
                  <br/>
				  
                  <h2 id="expression">Computing expression values for single cells</h2>
                  <br/>
                  <p>To use Monocle, you must first compute the expression of each gene in each cell for your experiment.  There are a number of ways to do this for RNA-Seq.  We recommend using Cufflinks, but you could also use RSEM, eXpress, Sailfish, or another tool for estimating gene and transcript expression levels from aligned reads.  Here, we'll show a simplified workflow for using TopHat and Cufflinks to estimate expression.  You can read more about how to use TopHat and Cufflinks to calculate expression <a href="http://www.nature.com/nprot/journal/v7/n3/full/nprot.2012.016.html">here</a>.
				  
				  <p>To estimate gene and transcript expression levels for single-cell RNA-Seq using TopHat and Cufflinks, you must have a file of RNA-Seq reads for each cell you captured. If you performed paired-end RNA-Seq, you should have two files for each cell. Depending on how the base calling was performed, the naming conventions for these files may differ.  In the examples below, we assume that each file follows the format:
				  <blockquote>
				  CELL_TXX_YYY.RZ.fastq.gz
				  </blockquote> 
				  Where XX is the time point at which the cell was collected in our experiment, YY is the well of the 96-well plate used during library prep, and Z is either 1 or 2 depending on whether we are looking at the left mate or the right mate in a paired end sequencing run.  So <tt>CELL_T24_A01.R1.fastq.gz</tt> means we are looking at the left mate file for a cell collected 24 hours into our experiment and which was prepped in well A01 of the 24-hour capture plate.
				  We begin by aligning each cell's reads separately, so we will have one BAM file for each cell. The commands below show how to run each cell's reads through TopHat.  These alignment commands can take a while, but they can be run in parallel if you have access to a compute cluster.  If so, contact your cluster administrator for more information on how to run TopHat in a cluster environment.
				  
                  <br/>
				  <br/>
				  <blockquote>
					  tophat -o CELL_T24_A01_thout -G GENCODE.gtf bowtie-hg19-idx CELL_T24_A01.R1.fastq.gz CELL_T24_A01.R2.fastq.gz <br/>
					  tophat -o CELL_T24_A02_thout -G GENCODE.gtf bowtie-hg19-idx CELL_T24_A02.R1.fastq.gz CELL_T24_A02.R2.fastq.gz <br/>
					  tophat -o CELL_T24_A03_thout -G GENCODE.gtf bowtie-hg19-idx CELL_T24_A03.R1.fastq.gz CELL_T24_A03.R2.fastq.gz <br/>
				  </blockquote>
				  <br/>
				  <p>The commands above show how to align the reads for each of three cells in the experiment.  You will need to run a similar command for each cell you wish to include in your analysis. These TopHat alignment commands are simplified for brevity - there are options to control the number of CPUs used by TopHat and otherwise control how TopHat aligns reads that you may want to explore on the <a href="http://tophat.cbcb.umd.edu/">TopHat</a> manual.  The key components of the above commands are:
				  <ol>
					  <li> the <tt>-o</tt> option, which sets the directory in which each cell's output will be written.
					  <li> the gene annotation file, specified with <tt>-G</tt>, which tells TopHat where to look for splice junctions.
					  <li> the Bowtie index for genome of your organism, in this case build hg19 of the human genome.
					  <li> the read files for each cell as mentioned above.
				  </ol>
				  
				  <p>When the commands finish, there will be a BAM file in each cell's TopHat output directory. For example, <tt>CELL_T24_A01_thout/accepted_hits.bam</tt> will contain the alignments for cell T24_A01.
				  
				  <p>Next, we will use Cufflinks to estimate gene expression levels for each cell in your study.
				  
                  <br/>
				  <br/>
				  <blockquote>
					  cuffquant -o CELL_T24_A01_cuffquant_out GENCODE.gtf CELL_T24_A01_thout/accepted_hits.bam <br/>
					  cuffquant -o CELL_T24_A02_cuffquant_out GENCODE.gtf CELL_T24_A02_thout/accepted_hits.bam <br/>
					  cuffquant -o CELL_T24_A03_cuffquant_out GENCODE.gtf CELL_T24_A03_thout/accepted_hits.bam <br/>
				  </blockquote>
				  <br/>
				  
				  <p>The commands above show how to convert aligned reads for each cell into gene expression values for that cell.  You will need to run a similar command for each cell you wish to include in your analysis. These commands are simplified for brevity - there are options to control the number of CPUs used by the cuffquant utility and otherwise control how cuffquant estimates expression that you may want to explore on the <a href="http://cufflinks.cbcb.umd.edu/">Cufflinks</a> manual.  The key components of the above commands are:
				  <ol>
					  <li> the <tt>-o</tt> option, which sets the directory in which each cell's output will be written.
					  <li> the gene annotation file, which tells cuffquant what the gene structures are in the genome.
					  <li> the BAM file containing the aligned reads.
				  </ol>
				  
				  <p>Next, you will need to merge the expression estimates into a single table for use with Monocle. You can do this with the following command:
					  <br/>
					  <br/>
					  <blockquote>
					  cuffnorm --use-sample-sheet \<br/>-o sc_expr_out GENCODE.gtf sample_sheet.txt
					  </blockquote>
					  <br/>
					  <p> The option <tt>--use-sample-sheet</tt> tells cuffnorm that it should look in the file <tt>sample_sheet.txt</tt> for the expression files, to make the above command simpler.  If you choose not to use a sample sheet, you will need to specify the expression files on the command line directly.  The sample sheet is a tab-delimited file that looks like this:
					  <br/>
					  <br/>
					  <blockquote>
						  sample_name     group<br/>
						  CELL_T24_A01_cuffquant_out/abundances.cxb  T24_A01<br/>
						  CELL_T24_A02_cuffquant_out/abundances.cxb  T24_A02<br/>
						  CELL_T24_A03_cuffquant_out/abundances.cxb  T24_A03<br/>
					  </blockquote>
				      <br/>
				  <p>Now we need to prepare files to use with Monocle from the Cufflinks package's output.  <a href="http://compbio.mit.edu/cummeRbund/">CummeRbund</a> provides an easy way to do this. You will need to <a href="http://www.bioconductor.org/packages/release/bioc/html/cummeRbund.html">install the CummeRbund package</a> through Bioconductor to do this.  Alternatively, see the Monocle vignette for instructions on how to prepare the input files if you wish to do so by other means.  The following R script loads the output from <tt>cuffnorm</tt>, converts it into a set of three tables, and writes these out as tab delimited files ready for use with Monocle. 
					  <br/>
					  <br/>
					  <blockquote>
					  > library(cummeRbund)<br/>
					  > cd <- readCufflinks("sc_expr_out")<br/>
					  > write.table(fpkmMatrix(genes(cd)), "expr_matrix.txt", quote=F, col.names=T, row.names=T, sep="\t")<br/>
					  > gene_ann <-annotation(genes(cd))<br/>
					  > row.names(gene_ann) <- gene_ann$gene_id<br/>
					  > write.table(gene_ann[,c("gene_short_name")]), "feature_data.txt", quote=F, col.names=T, row.names=T, sep="\t")<br/>
					  > cell_ann <-data.frame(row.names=samples(cd)$sample_id,
					  						cell_type="MyCellType")<br/>
					  > write.table(cell_ann, "pheno_data.txt", quote=F, col.names=T, row.names=T, sep="\t")<br/>
					  </blockquote>
					  <br/>
					  
					  Now, you are ready to load the expression data into Monocle and start analyzing your experiment.
					  <br/>
					  <br/>
                  <h2 id="monocle">Analyzing data with Monocle</h2>
                  <br/>
				  Monocle provides a number of tools you can use to analyze your single cell expression experiments. To get started, we must create a CellDataSet object.  You can do this with the commands below: 
				  <blockquote>
				  > library(monocle)<br/>
				  > sample_sheet <- read.delim("pheno_data.txt", row.names=1)<br/>
				  > gene_annotations <- read.delim("feature_data.txt", row.names=1)<br/>
				  > fpkm_matrix <- read.delim("expr_matrix.txt")<br/>
				  > pd <- new("AnnotatedDataFrame", data = sample_sheet)<br/>
				  > my_data <- new("CellDataSet", exprs = as.matrix(fpkm_matrix), phenoData = pd, featureData = fd)<br/>			  
			      </blockquote>
				  
				  Now, you have created an object named "my_data" that stores your single-cell expression data. This object is the central object in Monocle.  You will use it to identify differentially expressed genes and perform other analyses.  To see what Monocle can do for you and how to proceed, please have a look at the <a href="monocle.pdf">vignette (PDF)</a>.  Good luck!
				  
				  <br/>
				  
                  <br/>
                  <div id="footer">
                    <table cellspacing="15" width="100%">
                      <tbody>
                        <tr>
                          <td> This research was supported in part by
                            NIH grants R01-LM06845 and R01-GM083873, NSF
                            grant CCF-0347992 and the Miller Institute
                            for Basic Research in Science at UC
                            Berkeley. </td>
                          <td align="right"> Administrator: <a
                              href="mailto:cole@cs.umd.edu">Cole
                              Trapnell</a>. Design by <a
                              href="http://www.free-css-templates.com"
                              title="Design by David Herreman">David
                              Herreman</a> </td>
                        </tr>
                      </tbody>
                    </table>
                  </div>

                  <!-- Google analytics code -->
                  <script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
                  <script type="text/javascript">
try {
var pageTracker = _gat._getTracker("UA-6101038-2");
pageTracker._trackPageview();
} catch(err) {}</script>
                  <!-- End google analytics code --> </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </body>
</html>
